node/impl/full/eth.go

```diff
@@ -1270,6 +1325,53 @@ func (e *EthEventHandler) EthGetLogs(ctx context.Context, filterSpec *ethtypes.E
 	return ethFilterResultFromEvents(ctx, ces, e.SubManager.StateAPI)
 }
 
+func (e *EthEventHandler) waitForHeightProcessed(ctx context.Context, height abi.ChainEpoch) error {
+	ei := e.EventFilterManager.EventIndex
+	if height > e.Chain.GetHeaviestTipSet().Height() {
+		return xerrors.New("height is in the future")
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, eventReadTimeout)
+	defer cancel()
+
+	// if the height we're interested in has already been indexed -> there's nothing to do here
+	b, err := ei.IsHeightProcessed(ctx, uint64(height))
+	if err != nil {
+		return xerrors.Errorf("failed to check if event index has events for given height: %w", err)
+	}
+	if b {
+		return nil
+	}
+
+	// subscribe for updates to the event index
+	subCh, unSubscribeF := ei.SubscribeUpdates()
+	defer unSubscribeF()
+
+	// it could be that the event index was update while the subscription was being processed -> check if index has what we need now
+	b, err = ei.IsHeightProcessed(ctx, uint64(height))
+	if err != nil {
+		return xerrors.Errorf("failed to check if event index has events for given height: %w", err)
+	}
+	if b {
+		return nil
+	}
+
+	for {
+		select {
+		case <-subCh:
+			b, err = ei.IsHeightProcessed(ctx, uint64(height))
```

Comment:
why don't we just return the height on the object coming from the channel and check that rather than needing to do a round-trip through the db at each epoch that's processed? we could have thousands of these queued up waiting for a far-future epoch, each one doing a SELECT on each epoch while they wait

Reply:
oh, nevermind, `maxHeight > e.Chain.GetHeaviestTipSet().Height()` takes care of that concern